regression_em/model.py

```diff
@@ -1,18 +1,258 @@
+
+    def update_params(self, feat_mat: np.ndarray, responsibilities: np.ndarray, sample_weights) -> Tuple[RegParam, float]:
+        """Return fitted Logistic Regression params.
+        For detail, see eq.2 of [https://static.googleusercontent.com/media/research.google.com/ja//pubs/archive/46485.pdf].
+
+        :param feat_mat: Feature matrix to be used to learn responsibility.
+        :param responsibilities: Sequence of responsibilities calculated at E-step.
+        :param sample_weights: Sequence of Weights for treating imbalance dataset.
+        for positive data: 1/n-positive data
+        for negative data: 1/n-negative data
+        :return: Updated M-step params.
+        """
+        if self._alpha:
+            reg = Ridge(alpha=self._alpha)
+        else:
+            reg = LinearRegression()
+        reg.fit(feat_mat, self.calc_logits(responsibilities), sample_weights)
+        return reg.coef_, reg.intercept_
+
+    def _calc_log_likelihood(self, left_feat: np.ndarray, right_feat: np.ndarray,
+                             labels: np.ndarray) -> np.ndarray:
+        """Return log likelihood.
+        positive label: log(outcome_probs)
+        negative label: log(1-outcome_probs)
+
+        :param left_feat: Feature matrix to be used to learn left params.
+        :param right_feat: Feature matrix to be used to learn  parightrams.
+        :param labels: Sequence of boolean indicating each sample is positivei or negative.
+        :return: log_likelihood
+        """
+        # Calculate predicted probabilities of outcome.
+        outcome_probs = self.calc_probs(self.left_params[0], self.left_params[1], left_feat) * \
+            self.calc_probs(self.right_params[0], self.right_params[1], right_feat)
+
+        # Calculate log likelihood with positive samples only.
+        positive_sample_probs = outcome_probs[labels]  # Get list of probs whose sample labels are True.
+        positive_sample_probs[positive_sample_probs == 0] = self.epsilon  # Clip probs to avoid log(0).
+        positive_sample_log_lh = np.sum(np.log(positive_sample_probs))
+
+        # Apply the same procedure to negative ones.
+        negative_sample_probs = outcome_probs[~labels]
+        negative_sample_probs[negative_sample_probs == 1] = 1 - self.epsilon
+        negative_sample_log_lh = np.sum(np.log(1 - negative_sample_probs))
+
+        return positive_sample_log_lh + negative_sample_log_lh
+
+    # def fit(self, left_feat: np.ndarray, right_feat: np.ndarray, labels: np.ndarray) -> None:
+    def fit(self, X, y, index: int) -> None:
+        """Estimate regression EM params.
+
+        :param X: {array-like, sparse matrix} of shape (n_samples, n_left and right latent features)
+                  Feature matrix derived from concatenating left latent features with right latent features.
+        :param y: array-like of shape (n_samples,)
+                  Sequence of labels indicating each sample is positive or negative.
+        :param index: int
```

Comment:
Don't accept a boundary index here! A user can pass different index during training and predicting. It leads to inconsistency of feature shape.

Reply:
define split_index in `__init__` func.